src/system/arena_resource.hh

```diff
@@ -0,0 +1,210 @@
+    template <class T>
+    T* allocate_simd(size_t n_items) {
+        size_t bytes = sizeof(T) * n_items;
+        AlignedMemoryResource<Resource> a_res(this->resource_, simd_alignment_);
+        return (T*)a_res.do_allocate(bytes, a_res.alignment_);
+    }
+
+    // Reset allocated data
+    void reset() {
+        resource_.release();
+    }
+
+protected:
+    /// Override do_allocate to handle allocation logic
+    void* do_allocate(FMT_UNUSED size_t bytes, FMT_UNUSED size_t alignment) override {
+        // No-op
+        return nullptr;
+    }
+
+    /// Override do_deallocate (no-op for monotonic buffer)
+    void do_deallocate(FMT_UNUSED void* p, FMT_UNUSED size_t bytes, FMT_UNUSED size_t alignment) override {
+        // No-op
+    }
+
+    /// Override do_is_equal for memory resource comparison
+    bool do_is_equal(const std::pmr::memory_resource& other) const noexcept override {
+        return this == &other;
+    }
+
+private:
+    std::unique_ptr<char[]> buffer;
+    size_t buffer_size;
+    Resource resource_;
+    size_t simd_alignment_;
+};
+
+
+using AssemblyArena = ArenaResource<std::pmr::monotonic_buffer_resource>;
+using PatchArenaResource = ArenaResource<AssemblyArena>;
+
+
+/**
+ * Define vector allocated in Arena and aligned to SIMD size.
+ */
+template<class T>
+class ArenaVec {
+public:
+    /// Type definition
+	typedef Eigen::Matrix<T, Eigen::Dynamic, 1> VecData;
+
+	/// Default constructor, set invalid data pointer
+    ArenaVec()
+    : data_ptr_(nullptr), data_size_(0), arena_(nullptr) {}
+
+    /**
+     * Constructor. Set sizes and allocate data pointer
+     */
+    ArenaVec(size_t data_size, AssemblyArena &arena)
+    : data_ptr_(nullptr), data_size_(data_size), arena_(&arena) {
+        data_ptr_ = arena_->allocate_simd<T>( data_size_ );
+    }
+
+    /**
+     * Maps data pointer to Eigen Map of dimensions given data_size_ and returns it.
+     */
+    inline Eigen::Map<VecData> eigen_map() {
+        return Eigen::Map<VecData>(data_ptr_, data_size_, 1);
+    }
+
+    /// Return data pointer (development method)
+    T* data_ptr() {
+        return data_ptr_;
+    }
+
+    /// Getter for data_size_
+    inline size_t data_size() const {
+    	return data_size_;
+    }
+
+    inline T & operator()(std::size_t item) {
+        ASSERT_LT(item, data_size_);
+        return data_ptr_[item];
+    }
+
+    inline ArenaVec<T> operator+(const ArenaVec<T> &other) const {
+        ASSERT_EQ(data_size_, other.data_size());
+        ArenaVec<T> res(data_size_, *arena_);
+        Eigen::Map<VecData> result_map = res.eigen_map();
+        result_map = this->eigen_map() + other.eigen_map();
+        return res;
+    }
+
+    inline ArenaVec<T> operator*(T multi) const {
+        ArenaVec<T> res(data_size_, *arena_);
+        Eigen::Map<VecData> result_map = res.eigen_map();
+        result_map = this->eigen_map() * multi;
+        return res;
+    }
+
+protected:
+    T* data_ptr_;            ///< Pointer to data array
```

Comment:
ArenaVec musí mít default constructor s parametrem T, ten je potřeba si zapamatovat, protože v tu chvíli není známá arena.

Reply:
To odpovídá chybějícímu konstruktoru ArenaVec(T scalar).